agreement_dynamic/models/agreement_section.py

```diff
@@ -0,0 +1,258 @@
+        if (
+            all([view_id_tree == result["view_id"], view_type == "tree"])
+            or all([view_id_form == result["view_id"], view_type == "form"])
+        ) and self.env.context.get("is_locked"):
+            arch = etree.fromstring(result["arch"])
+            arch.attrib["create"] = "false"
+            arch.attrib["edit"] = "false"
+            arch.attrib["delete"] = "false"
+            result["arch"] = etree.tostring(arch)
+        return result
+
+    name = fields.Char("Name", required=True)
+    sequence = fields.Integer("Sequence", default=10)
+    content = fields.Html("Content")
+    dynamic_content = fields.Html(
+        compute="_compute_dynamic_content", string="Dynamic Content",
+    )
+    agreement_id = fields.Many2one("agreement", string="Agreement", ondelete="cascade")
+    resource_ref = fields.Reference(related="agreement_id.resource_ref")
+    res_id = fields.Integer(related="agreement_id.res_id")
+    resource_ref_model_id = fields.Many2one("ir.model", related="agreement_id.model_id")
+
+    # Dynamic field editor
+    field_id = fields.Many2one("ir.model.fields", string="Field")
+    sub_object_id = fields.Many2one("ir.model", string="Sub-model")
+    sub_model_object_field_id = fields.Many2one("ir.model.fields", string="Sub-field")
+    default_value = fields.Char("Default Value")
+    copyvalue = fields.Char("Placeholder Expression")
+    is_paragraph = fields.Boolean(
+        default=True, string="Paragraph", help="To highlight lines"
+    )
+    condition_python = fields.Text(
+        string="Python Condition", help="Condition for rendering section",
+    )
+    condition_domain = fields.Char(string="Domain Condition", default="[]")
+    condition_python_preview = fields.Char("Preview", compute="_compute_condition")
+    show_in_report = fields.Boolean(compute="_compute_condition")
+    model_id_model = fields.Char(
+        string="Model _description", related="agreement_id.model_id.model"
+    )
+
+    @api.onchange("field_id", "sub_model_object_field_id", "default_value")
+    def onchange_copyvalue(self):
+        self.sub_object_id = False
+        self.copyvalue = False
+        if self.field_id and not self.field_id.relation:
+            self.copyvalue = "${{object.{} or {}}}".format(
+                self.field_id.name, self._get_proper_default_value()
+            )
+            self.sub_model_object_field_id = False
+        if self.field_id and self.field_id.relation:
+            self.sub_object_id = self.env["ir.model"].search(
+                [("model", "=", self.field_id.relation)]
+            )[0]
+        if self.sub_model_object_field_id:
+            self.copyvalue = "${{object.{}.{} or {}}}".format(
+                self.field_id.name,
+                self.sub_model_object_field_id.name,
+                self._get_proper_default_value(),
+            )
+
+    @api.onchange("condition_python")
```

Comment:
What shall we do with `condition_domain` in this setup?

Reply:
Maybe I'm misssing something, but I don't see why `condition_domain` should be referenced at all in this computation function. The preview is for the Python part, and from what I understood, the `condition_domain` was only pulled in so that you could calculate `show_in_report` - if you're moving that to the render function, it's not needed here anymore.

If we still want to have a preview also for the `condition_domain`, we could have a separate preview field for that, but Odoo's "domain" widget already offers a nice preview - it shows the matched records and you can click on them to open.

Reply:
Maybe I missed something, because clearly I do not understand where `condition_domain` is needed then. I just removed it from the condition compute method, and now it appears nowhere at all, and influences nothing

Reply:
And then, based on the evaluation of `condition_python` and `condition_domain`, we have three options for the rendering:

1. Empty
2. The rendering result
3. A traceback, which we could render inside of a `<pre></pre>` block to make it be rendered as fixed-width text.

Reply:
If any of this sounds problematic just holler - to me it sounds pretty logical, but of course I could be missing something.

Reply:
OK, I probably misunderstood how the user story works. 
What I have understood is that we have a model, a record of that model, and a qweb template to render using that record. Condition_python can be then evaluated with that record. 
What should be evaluated by condition_domain then? And more importantly, how? That's what I am missing.

So when this is clear in my head, I'll do the following:

1. Create two separate methods, eval_python, eval_domain
2. In eval_python I will do what happens also now in the method that computes python_preview, at least a part of it
3. In eval_domain I will do something similar, when you explain to me how to do it
4. In dynamic_content I will check these conditions before rendering, and then show or not show the section

As for the case we have a problem in mako-rendering, you will see that I catch the exception, and the rendering_content becomes the traceback you mention above.